s3_storage/models/s3storage.py

```diff
@@ -50,64 +63,52 @@ def _get_storage_client(self):
 
     @api.model
     def _file_read(self, fname, bin_size=False):
+        if self._storage() != 's3':
+            return super(S3Attachment, self)._file_read(fname, bin_size)
         client, bucket = self._get_storage_client()
-        r = ''
+        res = ''
         try:
             if bin_size:
                 info = client.stat_object(bucket, fname)
-                r = human_size(info.size)
+                res = human_size(info.size)
             else:
                 response = client.get_object(bucket, fname)
-                r = response.read().encode('base64')
-        except ResponseError as e:
-            _logger.info("_read_file (s3) reading %s", fname, exc_info=True)
-        return r
+                res = response.read().encode('base64')
+        except ResponseError as errn:
+            LOGGER.info("_read_file (s3) reading %s", fname, exc_info=True)
+        return res
 
     @api.model
     def _file_write(self, value):
+        if self._storage() != 's3':
+            return super(S3Attachment, self)._file_write(value)
         client, bucket = self._get_storage_client()
         bin_value = value.decode('base64')
         fname, _ = self._get_path(bin_value)
         try:
             client.put_object(
                 bucket, fname, StringIO(bin_value), len(bin_value))
-        except ResponseError as e:
-            _logger.info("_file_write (s3) writing %s", fname, exc_info=True)
+        except ResponseError as errn:
+            LOGGER.info("_file_write (s3) writing %s", fname, exc_info=True)
         return fname
 
     @api.model
-    def _file_gc(self):
-        if self._storage() != 'file':
-            return
+    def _file_delete(self, fname):
+        if self._storage() != 's3':
+            return super(S3Attachment, self)._file_delete()
         client, bucket = self._get_storage_client()
 
-        # See comment in the original implementation
-        cr = self._cr
-        cr.commit()
-        self._cr.execute("LOCK ir_attachment IN SHARE MODE")
-
-        self._cr.execute("SELECT fname FROM s3_garbage")
-        fnames = [r[0] for r in cr.fetchall()]
-        kept = set()
+        self.env.cr.execute("""
+        SELECT COUNT(*) FROM ir_attachment WHERE store_fname = %s""", (fname,))
+        fnames = [r[0] for r in self.env.cr.fetchall()]
```

Comment:
Won't this result in `fnames` containing just the number of attachments? Also, since we're now removing one file at a time, we should probably use `client.remove_object()` instead, which takes only one filename.

Reply:
True. On the delete method I backported to the v9 method, i'm still testing this

Reply:
fixed on f98628b